### DISCUSSION TITLE
Upgrade from Java 20 to 21 broke multi ParameterResolvers?

### DISCUSSION THREAD

**Discussion by thatsIch** [junit-team/junit-framework#3556] [Q&A]
Hello community,

I use parameter resolvers to create preconditional objects for the tests like

```java
class RandomIntegerParameterResolver implements ParameterResolver {
  @Override
  public boolean supportsParameter(ParameterContext parameterContext, ExtensionContext extensionContext) throws ParameterResolutionException {
      return parameterContext.getParameter().getType() == Integer.class;
  }

  @Override
  public Object resolveParameter(ParameterContext parameterContext, 
    ExtensionContext extensionContext) throws ParameterResolutionException {
      return ThreadLocalRandom.current().nextInt(0, 10);
  }
}
```

Sometimes it requires having multiple `ParameterResolver`s per test which can be injected through `@ExtendWith`:

```java
@Test
@ExtendWith(RandomIntegerParameterResolver.class)
@ExtendWith(RandomDoubleParameterResolver.class)
void test(Integer i, Double d) {
    ...
}
```

But since Java 21 it broke for me in some projects.

# Stacktrace

```
org.junit.jupiter.api.extension.ParameterResolutionException: Failed to resolve parameter [some.domain.Object arg1] in method [void some.domain.ObjectServiceIT.all(some.other.Object,some.domain.Object)]: class java.util.HashMap cannot be cast to class some.domain.Object (java.util.HashMap is in module java.base of loader 'bootstrap'; some.domain.Object is in unnamed module of loader 'app')
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
Caused by: java.lang.ClassCastException: class java.util.HashMap cannot be cast to class some.domain.Object (java.util.HashMap is in module java.base of loader 'bootstrap'; some.domain.Object is in unnamed module of loader 'app')
	at some.domain.ObjectParameterResolver.resolveParameter(ObjectParameterResolver.java:35)
	at some.domain.ObjectParameterResolver.resolveParameter(ObjectParameterResolver.java:14)
	... 2 more
```

where `line 35` ist just the return like `return ...` which fails and `line 14` is the class start.

# Insights

* This worked without flaws with `JUnit 5.10.1` (currently the [latest version](https://mvnrepository.com/artifact/org.junit.jupiter/junit-jupiter-engine/5.10.1)) running on Java 20 and Ubuntu 22.04.
* But when I upgraded to Java 21 it broke.
* Running the same test on Windows had the same result.
* Parallel execution does not matter when executing the tests through Maven.
* It does not fail when executing within Eclipse.
* Tests which only use **one** `@ExtendWith` do not fail

The above example **sadly** works so I have not found a good reproducer.
I try to identify in which layer the bug is.

**Has somebody else encountered any problems like this?**

I am currently trying to setup a reproducer for this problem, but I wanted to reach out here in parallel.

Thanks for reading!

**Comment by marcphilipp** [comment]
@thatsIch Any updates on the reproducer?

**Reply by thatsIch** [reply]
I had not the time to investigate further (felt like jumping into a rabbit hole). I got some free days before the holidays, so I will try to work from the other side and try to strip down the project to reduce the problem to a reproducer.

Thanks for the reminder.
